Mention IPv6 and round-robin DNS situation

michielbdejong opened the pull request:
I'm guessing that IPv6 records are ignored by the Simple HTTPS and DVSNI challenge methods? The current text seems to assume "the host for the domain" is uniquely defined, which puts an incorrect security model in the reader's head.

I think it's wise to explicitly exclude validation over IPv6 for now, because an attacker who controls the DNS zone could add an AAAA record to it without breaking the live site for most users, and if letsencrypt would send the challenge to that AAAA address, it would not be testing the "real" live site in practice.

jsha commented:
I'm in favor of clarifying our use of round-robin and IPv6. I disagree with your rationale for excluding IPv6: If an attacker controls the DNS zone, it's game over. They don't need to muck around with AAAA records to get a cert. If they want to be careful not to take down the live site, they can add an A record pointing to a reverse proxy.

Could you instead word this to indicate that the server will connect to at least one of the hosts found in A and AAAA records, at its discretion?

michielbdejong commented:
OK! Good point. That also simplifies it.

I left the warning though, that "Simple HTTPS validation of IPv6-only domains may not be supported by all servers." because I estimate that that is what would be the case, unless you put "Servers SHOULD be able to challenge hosts that resolve for IPv6, even if they don't resolve for IPv4".

In the case of letsencrypt.org, would simple https validation of an IPv6-only domain work?

jsha commented:
Thanks! This sentence no longer reads quite right: "Although a domain may resolve to multiple IPv4 and IPv6 addresses, the server will connect to at least one of the hosts found in A and AAAA records, at its discretion."

We are planning to support IPv6 for letsencrypt.org, though it's possible it may not be ready in time for launch.

michielbdejong commented:
Yes, I changed 'Although' to 'As', and squashed the commits.

OK, if it may not be supported then it's probably appropriate to warn about that in the spec.
